include/boost/spirit/home/x3/support/traits/container_traits.hpp

```diff
@@ -155,27 +165,21 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
     template <typename Container, typename Enable = void>
     struct append_container
     {
-        // Not all containers have "reserve"
-        template <typename Container_>
-        static void reserve(Container_& c, std::size_t size) {}
-
-        template <typename T, typename Allocator>
-        static void reserve(std::vector<T, Allocator>& c, std::size_t size)
+        static void reserve(Container& c, std::size_t size, std::false_type)
         {
-            c.reserve(size);
+            // Not all containers have "reserve"
         }
-       
-        template <typename Container_, typename Iterator>
-        static void insert(Container_& c, Iterator first, Iterator last)
+
+        static void reserve(Container& c, std::size_t size, std::true_type)
         {
-            std::copy(first, last, std::inserter(c, c.end()));
+            c.reserve(size);
         }
 
         template <typename Iterator>
         static bool call(Container& c, Iterator first, Iterator last)
         {
-            reserve(c, c.size() + std::distance(first, last));
-            insert(c, first, last);
+        	reserve(c, c.size() + std::distance(first, last), std::integral_constant<bool, detail::has_reserve_method<Container>::value>());
+        	c.insert(c.end(), first, last);
```

Comment:
c.end() returns iterator, c.insert accept const_iterator. There is unneeded conversion.